pyro/infer/mcmc/nuts.py

```diff
@@ -283,11 +294,13 @@ def sample(self, trace):
 
                 rand = pyro.sample("rand_t={}_treedepth={}".format(self._t, tree_depth),
                                    dist.Uniform(torch.zeros(1), torch.ones(1)))
-                if rand < new_tree.size / tree_size:
+                if ((tree_size > 0) and (rand < new_tree.size / tree_size)) \
+                        or ((tree_size == 0) and (new_tree.size > 0)):
                     accepted = True
                     z = new_tree.z_proposal
 
-                if self._is_turning(z_left, r_left, z_right, r_right):  # stop doubling
+                r_sum += new_tree.r_sum
```

Comment:
what a bug!!! it took me a lot of time to detect it. We should never do shortcut assignment for tensor.

Reply:
Yeah, in place tensor ops can give non deterministic results, in many cases. Just curious - what was the cause of failure here?

Reply:
@neerajprad When tree_depth = 0, we create a new_tree with 1 element. Hence r_sum = r_left + r_right, but the shortcut assignment gives r_sum = r_left. Because of this, `is_turning` will return True (r_sum - r_left = 0), which makes NUTS not run at all or just run with 1 velocity verlet step.